diskqueue.go

```diff
@@ -395,6 +422,188 @@ func (d *diskQueue) readOne() ([]byte, error) {
+	}
+
+	var totalMessages int64
+	err = binary.Read(d.reader, binary.BigEndian, &totalMessages)
+	if err != nil {
+		return err
+	}
+
+	// update depth with the remaining number of messages
+	d.depth -= totalMessages - d.readMessages
+
+	// we have not finished reading this file
+	if d.readFileNum == d.nextReadFileNum {
+		d.nextReadFileNum++
+		d.nextReadPos = 0
+	}
+
+	d.moveToNextReadFile()
+
+	return nil
+}
+
+// walk through all of the files in the DiskQueue directory
+func (d *diskQueue) walkDiskQueueDir(fn func(os.FileInfo) error) error {
+	fileInfos, err := ioutil.ReadDir(d.dataPath)
+
+	if err != nil {
+		return err
+	}
+
+	for _, fileInfo := range fileInfos {
+		// only go through files and skip directories
+		if !fileInfo.IsDir() {
+			err = fn(fileInfo)
+			if err != nil {
+				return err
+			}
+		}
+	}
+
+	return nil
+}
+
+func (d *diskQueue) getAllBadFileInfo() ([]os.FileInfo, error) {
+	var badFileInfos []os.FileInfo
+
+	getAllBadFileInfo := func(fileInfo os.FileInfo) error {
+		// only accept "bad" files created by this DiskQueue object
+		if badFileNameRegexp.MatchString(fileInfo.Name()) {
+			badFileInfos = append(badFileInfos, fileInfo)
+		}
+
+		return nil
+	}
+
+	err := d.walkDiskQueueDir(getAllBadFileInfo)
+
+	return badFileInfos, err
+}
+
+// get the accurate total non-"bad" file size
+func (d *diskQueue) updateTotalDiskSpaceUsed() error {
+	d.totalDiskSpaceUsed = maxMetaDataFileSize
+
+	updateTotalDiskSpaceUsed := func(fileInfo os.FileInfo) error {
+		// only accept files created by this DiskQueue object
+		if fileNameRegexp.MatchString(fileInfo.Name()) || badFileNameRegexp.MatchString(fileInfo.Name()) {
+			d.totalDiskSpaceUsed += fileInfo.Size()
+		}
+
+		return nil
+	}
+
+	return d.walkDiskQueueDir(updateTotalDiskSpaceUsed)
+}
+
+func (d *diskQueue) freeDiskSpace(expectedBytesIncrease int64) error {
+	var err error
+	var badFileInfos []os.FileInfo
+
+	badFileInfos, err = d.getAllBadFileInfo()
+	if err != nil {
+		d.logf(ERROR, "DISKQUEUE(%s) failed to retrieve all .bad file info - %s", d.name, err)
+	}
+
+	// keep freeing up disk space until we have enough space to write this message
+	for d.totalDiskSpaceUsed+expectedBytesIncrease > d.maxBytesDiskSpace {
+		if len(badFileInfos) > 0 {
+			// check if a .bad file exists. If it does, delete that first
+			oldestBadFileInfo := badFileInfos[0]
+			badFileFilePath := path.Join(d.dataPath, oldestBadFileInfo.Name())
+
+			err = os.Remove(badFileFilePath)
+			if err != nil {
+				d.logf(ERROR, "DISKQUEUE(%s) failed to remove .bad file(%s) - %s", d.name, oldestBadFileInfo.Name(), err)
+				d.updateTotalDiskSpaceUsed()
+				return err
+			} else {
+				// recaclulate total bad files disk size to get most accurate info
+				d.totalDiskSpaceUsed -= oldestBadFileInfo.Size()
+				d.logf(INFO, "DISKQUEUE(%s) removed .bad file(%s) of size(%d bytes) to free up disk space", d.name, oldestBadFileInfo.Name(), oldestBadFileInfo.Size())
+			}
+
+			badFileInfos = badFileInfos[1:]
```

Comment:
Extract line 543-557 into a function.  Because `deleteBadFile` is not complicated comparing to other logics,   we could infer what it does from its name. So we could hide them into `deleteBadFile`. In this way, it gives readers extra time to focus on other complicated logics.
```go
			if err := d.deleteBadFile(badFileInfos[0]); err != nil {
				return err
			}
			badFileInfos = badFileInfos[1:]
```
```go
func (d *diskQueue) deleteBadFile(oldestBadFileInfo os.FileInfo) error {
	badFileFilePath := path.Join(d.dataPath, oldestBadFileInfo.Name())

	err := os.Remove(badFileFilePath)
	if err != nil {
		d.logf(ERROR, "DISKQUEUE(%s) failed to remove .bad file(%s) - %s", d.name, oldestBadFileInfo.Name(), err)
		d.updateTotalDiskSpaceUsed()
		return err
	} else {
		// recaclulate total bad files disk size to get most accurate info
		d.totalDiskSpaceUsed -= oldestBadFileInfo.Size()
		d.logf(INFO, "DISKQUEUE(%s) removed .bad file(%s) of size(%d bytes) to free up disk space", d.name, oldestBadFileInfo.Name(), oldestBadFileInfo.Size())
	}
	return nil
}
```

Reply:
SG

Reply:
Since I have a `removeReadFile` function is it ok if I name this function `removeBadFile` instead of `deleteBadFile`?

Reply:
Sure.